### ISSUE TITLE
Npm install bin conflict

### ISSUE THREAD

**Issue by heikki** [npm/npm#7130]
Running `npm install` fails randomly if two modules have bin with the same name.

Sample package.json:

``` json
{
  "name": "npm-install-bin-conflict",
  "dependencies": {
    "gulp": "^3.8.10",
    "gulp-cli": "^0.1.4"
  }
}
```

Output when error occurs:

``` sh
∴ npm install
npm WARN package.json npm-install-bin-conflict@ No description
npm WARN package.json npm-install-bin-conflict@ No repository field.
npm WARN package.json npm-install-bin-conflict@ No README data

> v8flags@1.0.8 install /Users/heikki/Desktop/bar/node_modules/gulp-cli/node_modules/v8flags
> node fetch.js


> v8flags@1.0.8 install /Users/heikki/Desktop/bar/node_modules/gulp/node_modules/v8flags
> node fetch.js

npm ERR! Darwin 14.0.0
npm ERR! argv "node" "/Users/heikki/.nvm-fish/v0.10.35/bin/npm" "install"
npm ERR! node v0.10.35
npm ERR! npm  v2.1.18
npm ERR! path ../gulp-cli/bin/gulp.js
npm ERR! code EEXIST
npm ERR! errno 47

npm ERR! EEXIST, symlink '../gulp-cli/bin/gulp.js'
File exists: ../gulp-cli/bin/gulp.js
Move it away, and try again. 

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/heikki/Desktop/bar/npm-debug.log
```

Original issue: https://github.com/gulpjs/gulp-cli/issues/7
Npm-debug.log: https://gist.github.com/heikki/eae172b8d1c5a97ed5ef#file-npm-debug-log

**Comment by heikki** [comment]
I just tried with `npm@2.2.0` and there I see fails only ~20% or 30% of the time. It was more obvious with 2.1.18 version.

**Comment by silkentrance** [comment]
Both packages define a "bin" : { gulp : ... } and as such will overwrite each other.
Perhaps including @phated in this discussion will help.

**Comment by heikki** [comment]
Scripts for reproducing the issue: https://github.com/heikki/npm-install-bin-conflicts

There's two differents errors (**ENOENT** & **EEXIST**) that depend on how npm install is run. First one happens on second run 100% of the time. Second one is random and the same as I reported above.

**ENOENT:**
```sh
∴ npm-install-bin-conflicts ./enoent.sh
install 1
install 2
npm ERR! Darwin 14.1.0
npm ERR! argv "node" "/Users/heikki/.nvm-fish/v0.10.36/bin/npm" "install" "gulp" "gulp-cli"
npm ERR! node v0.10.36
npm ERR! npm  v2.5.1
npm ERR! path /Users/heikki/Desktop/npm-install-bin-conflicts/node_modules/.bin/gulp
npm ERR! code ENOENT
npm ERR! errno 34

npm ERR! enoent ENOENT, unlink '/Users/heikki/Desktop/npm-install-bin-conflicts/node_modules/.bin/gulp'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/heikki/Desktop/npm-install-bin-conflicts/npm-debug.log
```

**EEXIST:**
```sh
∴ npm-install-bin-conflicts ./eexist.sh
install 1
install 2
install 3
install 4
install 5
install 6
install 7
install 8
install 9
npm ERR! Darwin 14.1.0
npm ERR! argv "node" "/Users/heikki/.nvm-fish/v0.10.36/bin/npm" "install" "gulp" "gulp-cli"
npm ERR! node v0.10.36
npm ERR! npm  v2.5.1
npm ERR! path ../gulp-cli/bin/gulp.js
npm ERR! code EEXIST
npm ERR! errno 47

npm ERR! EEXIST, symlink '../gulp-cli/bin/gulp.js'
File exists: ../gulp-cli/bin/gulp.js
Move it away, and try again.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/heikki/Desktop/npm-install-bin-conflicts/npm-debug.log
```

**Comment by heikki** [comment]
I noticed this issue isn't labeled as a bug yet. Should npm handle this case somehow?

Let me know so I can speculate further or move along. :smile:

**Comment by silkentrance** [comment]
FYI Running the same install command twice will result in NPM actually overwriting the existing symlink.

**Comment by phated** [comment]
It is strange that @othiym23 told us it was a bug in the gulp issue but tagged it as support and documentation on this issue.

This is a blocking issue on gulp 4.0 release, among other things.

**Comment by othiym23** [comment]
I'm still trying to figure out what the expected behavior is in cases where two packages want the same link. I misunderstood the issue at first. because I didn't notice that that was the underlying problem. The simplest solution is to *always* error when conflicting packages try to write the same link name, but that obviously won't work for gulp. The other obvious choice is last write wins, but we're trying to get away from nondeterminism in the installer. Bower prompts on conflicts, but npm would still need a deterministic solution for non-interactive installs (e.g. scripted deploys). What would work for gulp, @phated?

**Comment by phated** [comment]
In the gulp case, both are the same which allows people to have `gulp` as a devDep and still be able to use the gulp command in npm scripts, without the need for another dep, ala grunt and grunt-cli.  What does apt do if you have clashing bins?

**Comment by othiym23** [comment]
It's a policy violation, which is why node gets installed as `/usr/bin/nodejs` on Debian. There's also `/etc/alternatives`, which is a set of symbolic links that basically act as a pointer chain between a generic name (e.g. `editor` or `java`) and specific implementations. npm could do something similar, but it would require work from package developers to work effectively. I'm also not sure it helps gulp, because in this case it's literally the same script.

**Comment by silkentrance** [comment]
Not so much though for locally installed node_modules/[.]?bin symlinks or /usr/local/bin symlinks under control of npm.

**Comment by othiym23** [comment]
@silkentrance but that is the exact problem we have here -- two copies of the same script getting linked, for logical reasons, by two different modules, for the same application. The gulp team could easily resolve this by making the switch over from gulp < 4 to gulp 4 + gulp-cli a hard one, but that's user-hostile, and I don't like that idea any more than they do.

**Comment by silkentrance** [comment]
@othiym23 I proposed this one already to the peops over at gulp, namely @phated, they seem to be rather ignorant of the fact that they are doing something wrong.

However, running npm twice with the same arguments will actually do the trick, overwriting the existing symlinks, and which should equally fail. I'd call this to be the actual bug rather than the reported one.

**Comment by heikki** [comment]
@silkentrance Please don't bring your personal issues here. You're contributing unnecessary noise at the moment. 

The outcome of this issue should be whether and how npm handles the problem.

**Comment by silkentrance** [comment]
@heikki these are not personal issues but rather facts. i dunno where you peops get this from... *sigh*

**Comment by silkentrance** [comment]
@heikki it seems as if the peops over at gulp try to induce unwanted behaviour to existing, well-proven, tools in order for them to not rework their existing, non best practice based, mediocre solutions to problems that have long been solved. And for a special case such as gulp I do not see any benefit other than solving the actual problem

> However, running npm twice with the same arguments will actually do the trick, overwriting the existing symlinks, and which should equally fail. I'd call this to be the actual bug rather than the reported one.

**Comment by othiym23** [comment]
@silkentrance 
As I said up-thread, *always* erroring out when attempting to overwrite a symbolic link in `node_modules/.bin` is the simplest, most deterministic solution, but it doesn't make for a satisfying user experience, and it's going to be increasingly untenable over time. So far it's mostly luck and the convention that binary names follow package names that have kept this from being a larger issue, but once [scoped packages](https://docs.npmjs.com/misc/scope) come to npm's primary registry, I anticipate the need for a solution that doesn't require all script names to be unique over the entirety of the registry (which is, in effect, what that does).

Also, @phated and @heikki have been working this out with me in good faith on both this threads and the issue over on gulp's issue tracker, and they're very much aware of what they're asking for. This issue is both to work out a bug and an attempt by all of us to come up with a durable solution to the underlying problem. There's no reason to belittle them.

**Comment by silkentrance** [comment]
@othiym23 I never tried to belittle them, seriously. I just wanted them to realize that their special use case if far from established best practices. Perhaps they should use a pre-install hook instead, for both of their packages... but alas, the pre-install hook is already preoccupied by specialized node installation/build scripts.

As such, they should simply refrain from installing the same symlink twice and instead should break with their existing "API" and instead fall back to common best practice.

Again pointing out that

> However, running npm twice with the same arguments will actually do the trick, overwriting the existing symlinks, and which should equally fail. I'd call this to be the actual bug rather than the reported one.

**Comment by othiym23** [comment]
@silkentrance You've now included that four times, and all of us have seen it, acknowledged, and pointed out why that isn't the preferred solution. You don't need to point it out again.

**Comment by silkentrance** [comment]
@othiym23 It is the preferred solution as it would add a more deterministic behaviour to npm and not a second guess option.

**Comment by othiym23** [comment]
Let's enumerate what I believe the requirements to be around package binaries:

1. only one binary with the same name can exist at a time
2. packages can only specify one name for a binary in `package.json`
3. if two different packages configure a binary with the same name, that binary must end up linked to the same target after every install.

npm does not consistently enforce 3, which is the source of the bug. 3, as I've worded it, also doesn't prescribe a particular solution, which is by design. Figuring that out is what we're here to discuss.

We're here to figure out which solution makes for the smoothest user experience for both package maintainers and package consumers. Erroring out is *a* solution, but is not necessarily actionable for consumers, and is clunky for maintainers. gulp may be trying to square the circle here, but if so, they're not the only ones, and I consider their use case to be important enough that it warrants putting some time and creativity into coming up with a solution that doesn't force them (and others, in the future) into a corner.

However, @heikki / @phated, there is an issue here for gulp either way, because even if we do come up with a perfectly elegant solution we're all happy with, it's only going to work in newish versions of npm. You may be fine with mandating that `gulp@4` users upgrade to `npm@gulp-fixed-version`, but that's probably not going to stop a flood of issues from showing up in both npm's and gulp's issue trackers. If preventing that is the goal, it may be worth putting some time into coming up with a backwards-compatible solution that doesn't require everyone to be running `npm@latest`.

**Comment by silkentrance** [comment]
@othiym23 So you just basically say that the folks over at gulp should fix their process for now instead of relying on a provable future version of npm that solves all their issues, which might or might not come into place.

Considering point 3, that is, and which is is also the case of this issue, we are still running in circles.

And as I said before and repeating myself once again, rerunning the same npm install/update  command twice will replace existing symlink etc., as such the bug is different from what was reported, npm should never ever replace any existing files nor symlinks, regardless of how often it was invoked, unless it was --force'd to.

The above might be a first solution to the running-in-circles problem. Getting the peops over at gulp realize that their current solution is a no no and breaks with existing best practices, will eventually solve our problem.

**Comment by phated** [comment]
@othiym23 I agree, although it isn't the goal to have people installing both gulp and gulp-cli at the exact same time.  I foresee the upgrade path as two-fold:

1. User is perfectly happy using gulp 3 globally as it currently exists, but they want to upgrade to gulp 4, so they do `npm install -g gulp@latest` (or some variation there-of).
    * In this scenario, we've already prepped their CLI to handle old versions of gulp installed locally and everything works fine.
2.  User doesn't see the point in having the entirety of gulp installed globally, so they want to replace it with gulp-cli.  We urge them to do `npm uninstall -g gulp && npm install -g gulp-cli`.  If they don't heed our  instructions and just run `npm install -g gulp-cli`, we want the original (and non-backwards compatible) gulp binary to be removed and the gulp-cli to be linked in its place.
    * This user may go back and remove gulp whenever they do housecleaning, etc.

__Bonus__:
User is using gulp locally with npm-scripts, such as myself, and just upgrades to 4.0.  Nothing changes and the user doesn't need to add a `gulp-cli` dependency to their project.

Nowhere in our docs are we going to mention that someone should run `npm install gulp gulp-cli`, as that is unnecessary and I'm not sure what @heikki was experimenting with when he ran into the issue.  We just need to make sure that npm is consistent with how it handles the 2 upgrade paths and doesn't crash when trying to link/unlink the binaries.

tl;dr - I don't think there will be a flood of people hitting this issue, but I want a consistent condition to happen so we can easily guide people that do.

**Comment by heikki** [comment]
> If they don't heed our instructions and just run npm install -g gulp-cli, we want the original (and non-backwards compatible) gulp binary to be removed and the gulp-cli to be linked in its place.
This user may go back and remove gulp whenever they do housecleaning, etc.

This is a potential source for issues: `npm install -g gulp-cli && npm uninstall -g gulp`

There will be no gulp binary if user had an old version of npm.

**Comment by othiym23** [comment]
To me, this is yet more reinforcement of the idea that "last write wins" is not a useful solution. There is a clear preference for one `gulp` binary, and that's the one exported by `gulp-cli`. With that in mind, this feels more like an edge case, and less like something in want of a general solution.

How would you feel about the idea of having a postinstall script in `gulp@4` that checks to see if it's installed globally, and emits a warning saying that you should be removing it and globally installing `gulp-cli` instead? I think that may be easier for everybody than trying to work around two related packages exporting the same binary, even though it may lead to some short-term pain among people who don't ever read anything printed in the terminal ever (which includes as a subset people who depend on `*` in automated deploys).

That doesn't let npm off the hook, because requirement 3 above still stands. `npm@3` will be _more_deterministic, but I think it's still possible for the linking step at a given node in the dependency tree to end up performing links in different orders on different installs. Perhaps the solution is, indeed, to always error out when an attempt is made to create two links with the same name in `node_modules/.bin`.

**Comment by othiym23** [comment]
> This is a potential source for issues: `npm install -g gulp-cli && npm uninstall -g gulp`

That should actually go the other way around: `npm -g uninstall gulp && npm -g install gulp-cli`.

> There will be no gulp binary if user had an old version of npm.

Why not?

**Comment by silkentrance** [comment]
Nah they just need to copy existing grunt-cli/grunt package behaviour and they are good.

**Comment by phated** [comment]
@othiym23 what about linking the first one and then emitting a warning for the one trying to overwrite?

**Comment by heikki** [comment]
@othiym23 Yes, the problem free version is the other way round.

> Why not?

.. because it works that way now. :smile: 

```
∴ Desktop gulp -v
[22:00:18] CLI version 3.8.11

∴ Desktop npm install -g gulp-cli
/Users/heikki/.nvm-fish/v0.10.36/bin/gulp -> /Users/heikki/.nvm-fish/v0.10.36/lib/node_modules/gulp-cli/bin/gulp.js
gulp-cli@0.1.5 /Users/heikki/.nvm-fish/v0.10.36/lib/node_modules/gulp-cli
├── v8flags@2.0.2
├── pretty-hrtime@0.2.2
├── interpret@0.4.3
├── archy@1.0.0
├── minimist@1.1.0
├── semver@4.3.0
├── tildify@1.0.0 (user-home@1.1.1)
├── chalk@0.5.1 (escape-string-regexp@1.0.2, ansi-styles@1.1.0, supports-color@0.2.0, strip-ansi@0.3.0, has-ansi@0.1.0)
├── liftoff@2.0.1 (extend@1.3.0, flagged-respawn@0.3.1, resolve@1.0.0, findup-sync@0.2.1)
└── gulp-util@3.0.3 (array-differ@1.0.0, beeper@1.0.0, object-assign@2.0.0, array-uniq@1.0.2, replace-ext@0.0.1, lodash.reevaluate@3.0.1, lodash.reinterpolate@3.0.1, lodash.reescape@3.0.1, vinyl@0.4.6, through2@0.6.3, multipipe@0.1.2, lodash.template@3.2.0, dateformat@1.0.11)

∴ Desktop npm uninstall -g gulp
unbuild gulp@3.8.11

∴ Desktop gulp -v
fish: Unknown command 'gulp'
```

**Comment by othiym23** [comment]
I gotcha. Yes, that is indeed the bug, which just points out that the gulp team probably should come up with a solution that works regardless of the installed npm version, for the reasons I've already outlined.

> what about linking the first one and then emitting a warning for the one trying to overwrite?

This solution would work for the specific problem gulp is trying to solve right now, but doesn't generalize very far. Without some sort of stable ordering, "first one" could change from install to install / from host to host, which brings nondeterminism back into the picture. In the case of gulp, it's no biggie, but in a case where people have different programs with the same name (and, say, lifecycle scripts that depend on the named binaries), some scary things could happen.

**Comment by silkentrance** [comment]
Since most external modules/packages are installed using either npm [-g] install or via npm update, it seems that the --force option might do the trick?

**Comment by othiym23** [comment]
That's still not a good solution, because it doesn't address the underlying nondeterminism.

**Comment by silkentrance** [comment]
--force overrides any indeterminism

**Comment by othiym23** [comment]
That's not a solution, that's choosing to ignore the problem, which is not going to be adequate as this problem scales.

**Comment by Krinkle** [comment]
@othiym23 Per your request onhttps://github.com/jscs-dev/node-jscs/issues/787, here's mine: https://gist.github.com/Krinkle/69fbbf19eb31e4e56fa0

**Comment by Krinkle** [comment]
We re-images the servers and upgraded to npm v2.7.6. On a fresh build, it happened again (averaging one in every 7 builds or so).

```
Building remotely on integration-slave-trusty-1012 ( .. UbuntuTrusty) in workspace /mnt/jenkins-workspace/workspace/..
> rm -rf node_modules
> npm install
...
npm ERR! Linux 3.13.0-45-generic
npm ERR! argv "node" "/usr/local/bin/npm" "install"
npm ERR! node v0.10.25
npm ERR! npm  v2.7.6
npm ERR! path ../esprima-harmony-jscs/bin/esparse.js
npm ERR! code EEXIST
npm ERR! errno 47


npm ERR! EEXIST, symlink '../esprima-harmony-jscs/bin/esparse.js'
File exists: ../esprima-harmony-jscs/bin/esparse.js
Move it away, and try again. 
```

Full `npm-debug.log` at https://phabricator.wikimedia.org/P504#pastebin

**Comment by insin** [comment]
I have [a package](https://github.com/insin/eslint-config-jonnybuchanan) which acts as a proxy for its direct ESLint dependency by deliberately configuring its *own* `eslint` script in `bin`, which then finds the real `.bin/eslint` and calls it with some pre-configured arguments.

From my testing, when npm@3.5.0 installs my package, it installs ESLint into a nested `node_modules/` just like npm@2.x would have, which works great for my purposes, as nothing gets clobbered.

Is this behaviour reliable, and is a same-named `bin` in a parent and child dependency the same concern as this sibling dependency issue, or something different?

**Comment by ljharb** [comment]
@insin that seems very brittle - anyone else who includes your module, and also eslint, will cause a dedupe, such that your bin, and eslint's bin, may nondeterministically "win".

**Comment by nevir** [comment]
Another case where this is coming up pretty frequently is in npm v3 and projects that depend on `babel-cli@^6` and have a transitive dependency that depends on `babel@^5`.  Depending on install order, you either end up with a `babel-node` bin from babel v5 or v6

**Comment by Ajedi32** [comment]
Maybe I'm missing something here, but is there any reason we couldn't just namespace the bins and provide some kind of Debian-like `update-alternatives` system to allow users to determine which one gets to be the default in case of a conflict?
